### PR TITLE
Proposal: alternative entity factory API

### PR DESCRIPTION
**Pull request by TimMoore**
What do you think about this alternative?

### CODE REVIEW

File: persistence/play-cqrs/scaladsl/src/main/scala/play/scaladsl/cqrs/EntityFactory.scala [hunk 1]
```diff
@@ -19,15 +19,60 @@ import scala.reflect.ClassTag
...
+  def commandHandler: CommandHandler
+  def eventHandler: EventHandler
+  def tagger: Tagger[Event]
+}
+
 @ApiMayChange
-class EntityFactory[Command: ClassTag, Event, State](
-    name: String,
+class EntityFactory[Command <: ExpectingReply[_]: ClassTag, Event, State](
+    typeKey: EntityTypeKey[Command],
```
Comment (TimMoore) [inline persistence/play-cqrs/scaladsl/src/main/scala/play/scaladsl/cqrs/EntityFactory.scala:34]: The code in here is a bit rough... I was focusing on the end-user API in `AccountSampleCompileOnly`.

File: persistence/play-cqrs/scaladsl/src/main/scala/play/scaladsl/cqrs/CqrsComponents.scala [hunk 2]
```diff
@@ -3,33 +3,29 @@
...
 
-import akka.annotation.ApiMayChange
-
 @ApiMayChange
 trait CqrsComponents {
 
   def clusterSharding: ClusterSharding
 
-  final def createEntityFactory[Command: ClassTag, Event, State](
+  final def createEntityFactory[Command <: ExpectingReply[_]: ClassTag, Event, State](
```
Comment (patriknw) [inline persistence/play-cqrs/scaladsl/src/main/scala/play/scaladsl/cqrs/CqrsComponents.scala:18]: as mentioned by Renato I don't think we should strictly enforce this as the only way

File: persistence/play-cqrs/scaladsl/src/main/scala/play/scaladsl/cqrs/EntityFactory.scala [hunk 3]
```diff
@@ -19,15 +19,60 @@ import scala.reflect.ClassTag
...
 
+trait EntityDef[Command, Event, State] {
+  protected type CommandHandler = (State, Command) => ReplyEffect[Event, State]
+  protected type EventHandler = (State, Event) => State
+
+  def emptyState: State
+  def commandHandler: CommandHandler
+  def eventHandler: EventHandler
+  def tagger: Tagger[Event]
+}
```
Comment (patriknw) [inline persistence/play-cqrs/scaladsl/src/main/scala/play/scaladsl/cqrs/EntityFactory.scala:30]: what is this? It should be possible to define an EventSourcedBehavior in the same way in Akka and Play, to be able to move between those. This defeats that purpose. If we want some glue code for dependency injection and sharding initialization that can be fine, but not another way of defining the behavior.

**Review by patriknw — COMMENTED** [review]
skeptical

File: persistence/play-cqrs/scaladsl/src/main/scala/play/scaladsl/cqrs/EntityFactory.scala (hunk 3 above)
Comment (octonato) [reply persistence/play-cqrs/scaladsl/src/main/scala/play/scaladsl/cqrs/EntityFactory.scala:30]: This is only glueing the Tagger to the model and moving to an inheritance style over functional style. I think we can pass it. We can also create confusion with `EntityRef` and `EntityDef`. 

The Java API for EventSourcedBehavior is already using OO. To do the same will have to extend two classes. The behavior itself and then EntityDef without much added value.

**Comment by TimMoore** [comment]
Closing this, since the overall direction has evolved since last week.

**Pull request closed without merging**